### PR TITLE
Fix Intersection's of FiniteSet with symbolic elements 

### LINKED ISSUE #8217: Intersection(FiniteSet(x), FiniteSet(y)) shouldn't be EmptySet

### LINKED ISSUE #9536: Intersection's of FiniteSet with symbolic elements
Currently, either `FiniteSet` is able to evaluate intersection otherwise it, returns an `EmptySet()` .
 ```python
In [2] a = Symbol('a', real=True)

In [3]: FiniteSet(log(a)).intersect(S.Reals)
Out[3]: EmptySet()
```

It should rather return an unevaluated `Intersection` object.

### PR DESCRIPTION
**Pull request by aktech**
On Top of #8508
- [x] Fixes #8217
- [x] Fixes #9536 
- [x] Test for #9536 

TODO Failing Tests:
- [x] [sympy/solvers/tests/test_solveset.py](https://github.com/aktech/sympy/commit/a5ee04b5bd37df93e76c1918eee2fa297a9773fb)
- [x] [sympy/core/tests/test_wester.py](https://github.com/aktech/sympy/commit/e8e6a0bb9285c315e9bade2bcd10b954760d4965)
- [x] [sympy/stats/tests/test_rv.py](https://travis-ci.org/sympy/sympy/jobs/67831123)
- [x] [sympy/combinatorics/tests/test_partitions.py](https://travis-ci.org/sympy/sympy/jobs/67831130) [Though partitions can't be ordered (See issue #9608) ]

@hargup @flacjacket

### CODE REVIEW

**Comment by aktech** [comment]
I don't have enough idea about the failing tests.
There are some failing tests in `sympy.stats` .

**Comment by hargup** [comment]
These `intersection` or `union` should remain unevaluated only if its elements are symbols and the behaviour shouldn't change for elements which are fixed values.

```python
In [3]: f1 = FiniteSet(a, b) # Elements are symbols

In [4]: f2 = FiniteSet(b, c)

In [5]: f1.intersect(f2)
Out[5]: {b} U Intersection({a}, {b, c}) # Correct

In [6]: f1 = FiniteSet('a', 'b') # Elements are strings

In [7]: f2 = FiniteSet('b', 'c')

In [8]: f1.intersect(f2)
Out[8]: {b} U Intersection({a}, {b, c}) # Wrong; should be {b}



```

**Comment by aktech** [comment]
@hargup 
Strings doesn't have any special significance in `FiniteSet`, they behave identical to `Symbol` when mapped into a SymPy object.

```python
In [1]: from sympy import *

In [2]: sympy_object = sympify('string')

In [3]: type(sympy_object)
Out[3]: sympy.core.symbol.Symbol
```

See [here](https://github.com/sympy/sympy/blob/master/sympy/sets/sets.py#L1686) in `FiniteSet` definition.
Hence: `FiniteSet(a, b, c)` & `FiniteSet('a', 'b', 'c')` are equivalent.

```python
In [4]: f1 = FiniteSet('a', 'b', 'c')

In [5]: a, b, c = symbols('a, b, c')

In [6]: f2 = FiniteSet(a, b, c)

In [7]: f1 == f2
Out[7]: True

```

**Comment by aktech** [comment]
@hargup 
Please have a look at [this](https://github.com/aktech/sympy/commit/a5ee04b5bd37df93e76c1918eee2fa297a9773fb) commit. I am trying to fix failing test.

**Comment by aktech** [comment]
@hargup  @flacjacket 

```python
In [16]: x, y, z = symbols('x y z', integer=True)

In [19]: f1 = FiniteSet(x, y)

In [20]: f2 = FiniteSet(x, z)
```
### In Master:
```python
In [23]: f1.intersect(f2)
Out[23]: {x}
```

### This PR
```python
In [5]: f1.intersect(f2)
Out[5]: {x} U Intersection({y}, {x, z})
```

Clearly the behavior of FiniteSet in Master is non-acceptable, since in the above example `x, y, z` are integer symbols, so they can be any integer, but in the *Master* , they are assumed to be distinct, which is wrong.
There are such failing tests in `test_wester.py` [here](https://github.com/sympy/sympy/blob/master/sympy/core/tests/test_wester.py#L74), which should probably be corrected?

**Comment by jksuom** [comment]
More explicitly ``f1.intersect(f2)`` could be ``Piecewise(({x, y}, y == z), ({x}, True))``, but maybe that would be hard to implement in general.

**Comment by Arnab411** [comment]
GgGg

-----Original Message-----

**Comment by aktech** [comment]
TODO Failing Tests:
* [sympy/stats/tests/test_rv.py](https://travis-ci.org/sympy/sympy/jobs/67831123)
* [sympy/combinatorics/tests/test_partitions.py](https://travis-ci.org/sympy/sympy/jobs/67831130)

**Comment by jksuom** [comment]
``test_given`` fails because it tests the stochastic independence of variables in the construction of ``B`` by simply comparing their symbols ([``pspace_independent``](https://github.com/sympy/sympy/blob/master/sympy/stats/rv.py#L1042))

A quick-and-dirty solution could be adding a keyword to ``intersect`` telling if to fall back to the old behaviour. A cleaner solution might be changing all ``symbols`` to Python sets (and modifying ``intersect`` accordingly).

**Comment by aktech** [comment]
https://github.com/aktech/sympy/commit/e8e6a0bb9285c315e9bade2bcd10b954760d4965

**Comment by aktech** [comment]
> A cleaner solution might be changing all symbols to Python sets (and modifying intersect accordingly).

Is there a way to represent cases like [this](https://github.com/sympy/sympy/blob/master/sympy/stats/tests/test_finite_rv.py#L116)?

**Comment by jksuom** [comment]
I'm afraid I don't quite understand the issue. Changing the type of probability space [``symbols``](https://github.com/sympy/sympy/blob/master/sympy/stats/rv.py#L38) (i.e. the set of variable names) should not have any consequences here.

**Comment by aktech** [comment]
@jksuom 
Sorry I misinterpreted your last comment, It's perfectly fine. See changes here: https://github.com/sympy/sympy/commit/ae5fc1e9554a77fe989194928be111b42b9c4c85

**Comment by jksuom** [comment]
Looks good to me. But maybe the type of other probability space symbols should also be changed (e.g. product space).

**Comment by aktech** [comment]
@jksuom 
I just noticed that, we can't change the type of `symbols` to python `sets` for returning a class object in sympy, since all the `args` (obj.args) should be an instance of `Basic` in sympy. (otherwise `sympy/core/tests/test_args.py` would fail as [here](https://travis-ci.org/sympy/sympy/jobs/68607271).)
So we should only change the `@property` symbols for internal use.  See changes [here](https://github.com/aktech/sympy/commit/4129c66f7b26baac327fcec673a1398b4fe1a55a).

**Comment by jksuom** [comment]
That looks promising.

**Comment by jksuom** [comment]
Another possibility, with no effect on the interface, would be just turning ``a_symbols`` and ``b_symbols`` to Python sets in ``pspace_independent``.

**Comment by aktech** [comment]
It sounds good. Made changes here: https://github.com/aktech/sympy/commit/c68c9da985857bd9e61de36b192b32fa80e18cc9

File: sympy/sets/sets.py [hunk 1]
```diff
@@ -1374,12 +1374,15 @@ def flatten(arg):
         if len(args) == 0:
             raise TypeError("Intersection expected at least one argument")
 
+        try:
+            args = list(ordered(args, Set._infimum_key))
+        except TypeError:
```
Comment (hargup) [inline sympy/sets/sets.py]: Which scenarios raise this error?

File: sympy/sets/sets.py [hunk 2]
```diff
@@ -1431,8 +1434,14 @@ def reduce(args):
         # all other sets in the intersection
         for s in args:
             if s.is_FiniteSet:
-                return s.func(*[x for x in s
-                    if all(other.contains(x) == True for other in args)])
+                args = [a for a in args if a != s]
```
Comment (hargup) [inline sympy/sets/sets.py]: better name: `other_args`

File: sympy/sets/sets.py [hunk 3]
```diff
@@ -1431,8 +1434,14 @@ def reduce(args):
...
+                res = s.func(*[x for x in s
```
Comment (hargup) [inline sympy/sets/sets.py]: `res = FiniteSet(*[...])` will be more readable, it also goes with the Zen of Python [#1] `Explicit is better than Implicit`

[#1]: https://www.python.org/dev/peps/pep-0020/

File: sympy/sets/tests/test_sets.py [hunk 4]
```diff
@@ -238,7 +238,15 @@ def test_intersect():
     assert Union(Interval(0, 1), Interval(2, 3)).intersect(S.EmptySet) == \
         S.EmptySet
     assert Union(Interval(0, 5), FiniteSet('ham')).intersect(FiniteSet(2, 3, 4, 5, 6)) == \
-        FiniteSet(2, 3, 4, 5)
+        Union(FiniteSet(2, 3, 4, 5),
+              Intersection(FiniteSet(6), Union(Interval(0, 5),
```
Comment (hargup) [inline sympy/sets/tests/test_sets.py]: I guess you can ignore PEP8 here, this line break looks ugly.

File: sympy/sets/sets.py (hunk 1 above)
Comment (aktech) [inline sympy/sets/sets.py]: This one: https://travis-ci.org/sympy/sympy/jobs/67831130

**Comment by aktech** [comment]
@jksuom  @hargup  is there anything else to be addressed?

**Comment by hargup** [comment]
+1

**Comment by aktech** [comment]
> can give me self and other sets on which is_superset function is called?

It can be simulated by the following:
```python
from sympy import *
from sympy.combinatorics import Partition
p1 = Partition([1, 2], [3, 4])
p2 = Partition([4], [1, 2, 3])
print p2.is_proper_superset(p1)
```

**Comment by jksuom** [comment]
The ``stats`` part is fine, but I didn't look into test_partitions. I understood that it was fixed. +1

**Comment by hargup** [comment]
+1

**Pull request merged**